temporal-test-server/src/test/java/io/temporal/testserver/functional/NexusWorkflowTest.java

```diff
@@ -54,432 +54,617 @@
+      callerTask = pollWorkflowTask();
+      testWorkflowRule.assertHistoryEvent(
+          callerExecution.getWorkflowId(), EventType.EVENT_TYPE_NEXUS_OPERATION_COMPLETED);
+      completeWorkflow(callerTask.getTaskToken());
+    } catch (Exception e) {
+      Assert.fail(e.getMessage());
+    } finally {
+      nexusPoller.cancel(true);
+    }
+  }
+
+  @Test
+  public void testNexusOperationCancellation() {
+    String operationId = UUID.randomUUID().toString();
+    CompletableFuture<ByteString> nexusPoller =
+        pollNexusTask().thenCompose(task -> completeNexusTask(task.getTaskToken(), operationId));
+
+    try {
+      WorkflowStub stub = newWorkflowStub("TestNexusOperationTimeoutAfterCancelWorkflow");
+      WorkflowExecution execution = stub.start();
+
+      // Get first WFT and respond with ScheduleNexusOperation command
+      PollWorkflowTaskQueueResponse pollResp = pollWorkflowTask();
+      completeWorkflowTask(
+          pollResp.getTaskToken(),
+          newScheduleOperationCommand(
+              defaultScheduleOperationAttributes()
+                  .setScheduleToCloseTimeout(Durations.fromSeconds(5))));
+      testWorkflowRule.assertHistoryEvent(
+          execution.getWorkflowId(), EventType.EVENT_TYPE_NEXUS_OPERATION_SCHEDULED);
+
+      // Wait for operation to be started
+      nexusPoller.get();
+
+      // Poll and verify started event is recorded and triggers workflow progress
+      pollResp = pollWorkflowTask();
+      testWorkflowRule.assertHistoryEvent(
+          execution.getWorkflowId(), EventType.EVENT_TYPE_NEXUS_OPERATION_STARTED);
+      List<HistoryEvent> events =
+          testWorkflowRule.getHistoryEvents(
+              execution.getWorkflowId(), EventType.EVENT_TYPE_NEXUS_OPERATION_SCHEDULED);
+      Assert.assertEquals(1, events.size());
+
+      // Cancel operation
+      HistoryEvent scheduledEvent = events.get(0);
+      Command cancelCmd =
+          Command.newBuilder()
+              .setCommandType(CommandType.COMMAND_TYPE_REQUEST_CANCEL_NEXUS_OPERATION)
+              .setRequestCancelNexusOperationCommandAttributes(
+                  RequestCancelNexusOperationCommandAttributes.newBuilder()
+                      .setScheduledEventId(scheduledEvent.getEventId()))
+              .build();
+      completeWorkflowTask(pollResp.getTaskToken(), cancelCmd);
+
+      // Poll for and complete cancellation task
+      pollNexusTask()
+          .thenCompose(
+              task ->
+                  completeNexusTask(
+                      task.getTaskToken(),
+                      Response.newBuilder()
+                          .setCancelOperation(CancelOperationResponse.getDefaultInstance())
+                          .build()))
+          .get();
+
+      // Poll to verify cancellation is recorded and triggers workflow progress.
+      pollResp = pollWorkflowTask();
+      testWorkflowRule.assertHistoryEvent(
```

Comment:
Please validate the failure here has the same structure as the one in the real server.

Reply:
+1